src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,469 @@
+    {
+        private static unsafe Vector128<float> Load1(float* src, int* idx)
+        {
+            return Sse.SetScalarVector128(src[idx[0]]);
+        }
+
+        private static unsafe Vector128<float> Load4(float* src, int* idx)
+        {
+            return Sse.SetVector128(src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);
+        }
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            // The control byte shuffles the four 32-bit floats of x: ABCD -> BCDA.
+            return Sse.Shuffle(x, x, 0x39);
+        }
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            // The control byte shuffles the four 32-bit floats of x: ABCD -> DABC.
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
+            x = Rotate(x);
+            Sse.StoreScalar(dst + idx[1], x);
+            x = Rotate(x);
+            Sse.StoreScalar(dst + idx[2], x);
+            x = Rotate(x);
+            Sse.StoreScalar(dst + idx[3], x);
+        }
+
+        private static unsafe Vector128<float> VectorSum(Vector128<float> vector)
+        {
+            if (Sse3.IsSupported)
+            {
+                Vector128<float> tmp = Sse3.HorizontalAdd(vector, vector);
+                return Sse3.HorizontalAdd(tmp, tmp);
+            }
+            else
+            {
+                // SSE3 is not supported.
+                Vector128<float> tmp = Sse.Add(vector, Sse.MoveHighToLow(vector, vector));
+                // The control byte shuffles the four 32-bit floats of tmp: ABCD -> BADC.
+                return Sse.Add(tmp, Sse.Shuffle(tmp, tmp, 0xb1));
+            }
+        }
+
+        internal static unsafe void ScaleU(float scale, Span<float> dst)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            fixed (float* pdst = dst)
+            {
+                float* pDstCurrent = pdst;
+                float* pEnd = pdst + dst.Length;
+
+                while (pDstCurrent + 4 <= pEnd)
+                {
+                    Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                    dstVector = Sse.Multiply(scaleVector, dstVector);
+                    Sse.Store(pDstCurrent, dstVector);
+
+                    pDstCurrent += 4;
+                }
+
+                while (pDstCurrent < pEnd)
+                {
+                    Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                    dstVector = Sse.MultiplyScalar(scaleVector, dstVector);
+                    Sse.StoreScalar(pDstCurrent, dstVector);
+
+                    pDstCurrent++;
+                }
+            }
+        }
+
+        internal static unsafe void AddScaleU(float scale, Span<float> src, Span<float> dst)
+        {
```

Comment:
(minor) it may make sense to add some `Debug.Asserts` to check the src and dst Lengths match.  However, these are internal functions that are only called from functions that guarantee the arguments are checked, so I don't believe it is a blocking issue. It just may be some nice documentation on the expectations of these methods.  And in case they get new callsites in the future.